### ISSUE TITLE
Releasing DNS version 3.0.0

### ISSUE THREAD

**Issue by kazu-yamamoto** [kazu-yamamoto/dns#88]
I'm preparing DNS version 3.0.0.
Please check changelog:

https://github.com/kazu-yamamoto/dns/blob/master/Changelog.md

If nobody will raise objections, I will release version 3.0.0 after merging one security patch.

@themattchan @vdukhovni @adinapoli-iohk @themattchan @RyanGlScott @cleverca22 @erikd

**Comment by vdukhovni** [comment]
I have one long-standing issue that perhaps should be addressed before 3.0.0:

* Re-use of UDP sockets for anything other than retries of the same query is wrong.

This is because a delayed reply may come back after a timeout after we've already moved on to making a new query, and given sufficient retries, and possibly more timeouts on the next query, we can get out of sync with the nameserver, and fail to correlate requests and replies.

To fix this the resolver object should just a list of pre-resolved addresses, *NOT* connected sockets, and the sockets should be created just-in-time when looping over the nameserver list, with each socket closed after its sequence of retries or on success.

This has the added benefit of reducing file-descriptor pressure.  Each thread doing DNS will have at most one DNS query socket at a time, and only temporarily, instead of having multiple sockets open, many of which go unused, potentially for a long time if the same resolver object is used repeatedly.

This would also simplify the TCP code, as it would no longer need to grab the peer address from the UDP socket via getsockname(), and would just use the same address used to create the UDP socket on the fly.

At some point we could introduce a parallel queries on timeout, and there could again be concurrent open sockets, but that's for later.  The above should be enough for 3.0.0.  Any chance you're willing to implement this?

**Comment by adinapoli-iohk** [comment]
No objection on my side, so I will let the other folks comment 😉

**Comment by kazu-yamamoto** [comment]
@vdukhovni I don't have to hurry up and can wait. I thought I should release version 3.0.0 because I handled all PRs.

Unfortunately, I cannot take much time to maintain my Haskell libraries for the next 6 months (October to  March) because my mission in our company will change. I will come back in the next April.

> To fix this the resolver object should just a list of pre-resolved addresses, NOT connected sockets, and the sockets should be created just-in-time when looping over the nameserver list, with each socket closed after its sequence of retries or on success.

I think this idea is good. But I worry that creating socket might take a global lock in the kernel inside thus introduce heavy contention. We need to check.

**Comment by Gabriel439** [comment]
@kazu-yamamoto: If you will be busy in the next 6 months is there somebody else who can accept pull requests and publish releases during that time?

**Comment by vdukhovni** [comment]
> I think this idea is good. But I worry that creating socket might take a global lock in the kernel inside thus introduce heavy contention. We need to check.

It is true that binding a socket requires an unused port, and so some sort of mutual exclusion is needed.  Generally, the latency and cost of the actual DNS queries dominates the cost of creating a new bound socket.  If this cost was an issue, then UDP sockets could be re-used, provided they missed a reply.  Mind you, re-using the same socket increases opportunities for DNS forgery if the resolver is remote. DNS query source ports are supposed to be crypto-random to reduce opportunities for Kaminsky attacks.  Of course that's a bigger issue between iterative resolvers and the world than between stub resolvers and the iterator, especially if the stub resolver is talking to a resolver on the loopback interface, but still long-lived fixed-port clients carry some risk.

In my DANE survey I currently open a new resolver from the resolvSeed for each query.  The code does >1,000 DNS queries per second even while communicating with many diverse domains, some of which could be slow to respond.  I could go faster at greater concurrency, but some TLDs seem to impose rate limits that I would then be likely to cross.

I don't know what source of throughput you're looking for, but I would expect that the kernel's socket pool is not critical for the vast majority of applications.  Of course if one is querying repeatedly for the same DNS names, that are already in a local cache, then throughput might be more sensitive to small local costs, because the "remote" cost becomes small, but this assumes a truly local (e.g. loopback) cache, as otherwise the queries still need to go out on the network and come back...

My take is that correctness is for now more important than squeezing out the extra performance.

**Comment by vdukhovni** [comment]
> If you will be busy in the next 6 months is there somebody else who can accept pull requests and publish releases during that time?

Would running off your own fork in the mean-time meet your needs, or do you need others to be able to build your code against their own Network.DNS from hackage?

When I wanted a "bleeding edge" Network.DNS for my code, I added "dns" as a git submodule to my project, and that way, it is always built against a Network.DNS of my choice, which is at times my own fork.  See: http://github.com/vdukhovni/danecheck, which has "dns", "optparse-applicative" and "idna-hs" as submodules.

So long as @kazu-yamamoto ultimately has time to review accumulated patches, I am not too worried about the latency of features going to the upstream master.

**Comment by Gabriel439** [comment]
We have the ability to deploy off a Git revision (and that is what we do for `dns`).  However, I was thinking more of users who are using the `dns` library off Hackage as one dependency among many and aren't even aware that there are other branches that they should be using.  Publishing a blessed release early and often helps users who don't have as much time to devote to these sorts of things

**Comment by kazu-yamamoto** [comment]
> I don't know what source of throughput you're looking for, but I would expect that the kernel's socket pool is not critical for the vast majority of applications. Of course if one is querying repeatedly for the same DNS names, that are already in a local cache, then throughput might be more sensitive to small local costs, because the "remote" cost becomes small, but this assumes a truly local (e.g. loopback) cache, as otherwise the queries still need to go out on the network and come back...

This library is used in https://bazqux.com/
I want to keep the performance which bazqux requires.

**Comment by kazu-yamamoto** [comment]
Cc: @vshabanov

**Comment by kazu-yamamoto** [comment]
I would like to propose:
- We will release version 3.0.0 soon because this provides a feature to detect the default DNS servers on Windows.
- I will review and merge small RPs and release version 3.0.x, if necessary, for the next 6 months.
- During this period, somebody can develop big features and I will be able to review them or develop them by myself in the next April.

What do you think, guys?

**Comment by vshabanov** [comment]
I'd like to add few user facing things:
* checks for invalid domain names so there is no DNS request for "bad/domain!name"
* check for IP addresses (no request for "127.0.0.1" domain but just return its address, IPv6 would be nice too)
* function to get human readable error description (kind of strerror)
* catching exceptions in socket I/O (for example when socket could not be opened due to local BIND not running)

Here are utility functions that implement above functionality that I'm using in BazQux Reader:
https://gist.github.com/vshabanov/04b394e1cddad942c67357ad287f1552

Would be nice if they (or their more optimized/correct/beautiful versions) will be moved into 'dns'.

**Comment by vdukhovni** [comment]
> @vshabanov I'd like to add few user facing things:
>
> checks for invalid domain names so there is no DNS request for "bad/domain!name"

It is a common misconception that such names are not valid DNS names, but in fact DNS supports all octet strings in queries (https://tools.ietf.org/html/rfc2181#section-11) provided:
* No label is longer than 63 bytes
* The total (uncompressed) length of the wire-form of a domain name is not longer than 255 bytes
The above example is an invalid *hostname*, but it is a valid DNS label and DNS libraries need to support the full set of valid DNS names, not just hostnames. It is not clear how much effort we should expend to check the syntax of hostnames. One might also argue that the DNS library should not emit queries for ".onion" names, but am not convinced such checks are needed.

> check for IP addresses (no request for "127.0.0.1" domain but just return its address, IPv6 would be nice too)

This is a controversial feature.  It can hide the fact that such names are not, in fact, valid DNS names.  My own code would now have to start explicitly filtering out invalid address-valued MX hostnames, which previously would just not resolve.

> function to get human readable error description (kind of strerror)

DNS exceptions have "Show" instances, not sure what you're getting at.

> catching exceptions in socket I/O (for example when socket could not be opened due to local BIND not running)

This is already implemented.

**Comment by kazu-yamamoto** [comment]
> I am skeptical that the kernel overhead of opening sockets is a bottleneck for this application. Are there measurements that show this to be the case? Does the application use the same resolver object for all DNS queries? That risks getting out of sync with the nameserver and makes Kaminsky-style attacks easier. My intuition is that opening a new socket for each query (and its retransmissions if any) will be an improvement in reliability, and will make no noticeable difference in performance.

@vdukhovni Let's implement it and ask @vshabanov to test it.

**Comment by kazu-yamamoto** [comment]
> Sure. If I don't get to it before release, I'd like to think that lazy socket creation is a small RP, and that it could go in during that time (just a minor restructuring of the query loop and the internal resolver state).

Thanks.

> Sounds fair to me. IIRC EDNS0 support is not yet fully implemented, if so, it would be good to have that available.

Only the encoder/decoder are available. Nobody use them at this moment.

**Comment by kazu-yamamoto** [comment]
> - checks for invalid domain names so there is no DNS request for "bad/domain!name"
> - check for IP addresses (no request for "127.0.0.1" domain but just return its address, IPv6 would be nice too)

@vshabanov It would be nice to include such validation functions to `Network.DNS.Utils` so that users can combine them with lookup functions by themselves.

**Comment by kazu-yamamoto** [comment]
Such validation functions should have doctest cases to show their behavior.

**Comment by vshabanov** [comment]
> @vdukhovni Let's implement it and ask @vshabanov to test it.

OK. Although I'm using local BIND so I don't think there will be any major difference whether socket is opened again for each request or reused.

> @vshabanov It would be nice to include such validation functions to Network.DNS.Utils so that users can combine them with lookup functions by themselves.

Yes. Since it turned out that DNS could resolve almost anything invalid hostname and IP address hostname detection should be separate utility functions. But maybe 255 bytes total and 63 bytes per label check should live in core package since such names are invalid.

>> function to get human readable error description (kind of strerror)

> DNS exceptions have "Show" instances, not sure what you're getting at.

It's better to show end user "No such domain" than "NameError" and to not add new error strings to my code when dns library is updated.

**Comment by vdukhovni** [comment]
> > > @vshabanov function to get human readable error description (kind of strerror)

> > DNS exceptions have "Show" instances, not sure what you're getting at.

> @vshabanov  It's better to show end user "No such domain" than "NameError" and to not add new error strings to my code when dns library is updated.

Perhaps some user-friendly strings belong in a separate `displayDNSError` function, rather the Show instance of DNSError?  Does Haskell have support for localized error messages? It is not obvious that just English strings are sufficient in this context.

Your application is unlikely to encounter anything outside of:
```
   NameError                   -- The domain name does not exist
   ServFail                    -- The DNS lookup failed
   RetryLimitExceeded          -- All lookup attempts failed or timed out
   TimeoutExpired              -- (I don't see this one any more, see above)
   NetworkFailure  IOException -- Trouble reaching the configured resolver(s)
```
For the last one you may want to tease apart the IOException, see https://hackage.haskell.org/package/lens-4.15.4/docs/System-IO-Error-Lens.html, which provides access to `description` (aka `ioe_description` from `GHC.IO.Exception`, but sadly not exported via `System.IO.Error`)

The remaining error conditions are more typical of what the iterative resolver sees from authoritative resolvers, than what application stub resolvers (such as `Network.DNS`) see from from their iterative resolver, which typically hides the majority of error conditions behind `ServFail`.

Also keep in mind that you will sometimes get an empty list of answers (NODATA in DNS parlance), in which case `Network.DNS` will not return an error, though perhaps in some cases it should.  There is little practical difference between a name that exists, but has no addresses, and a name that does not exist when doing name -> address resolution.  If the application has a fallback query to try on the same name (try `AAAA` after trying `A`, or try `A` after trying `MX`) then it is useful to distinguish between empty replies and NameError conditions, but if not, then the distinction between NODATA and NXDomain is little interest and the two can be presented to the user identically ("address lookup failed", ...)

**Comment by kazu-yamamoto** [comment]
> Yes. Since it turned out that DNS could resolve almost anything invalid hostname and IP address hostname detection should be separate utility functions. But maybe 255 bytes total and 63 bytes per label check should live in core package since such names are invalid.

Agreed. A PR is welcome!

**Comment by kazu-yamamoto** [comment]
We should fix #90 and #91 before the release.

**Comment by kazu-yamamoto** [comment]
Though #90 and #91 have not been resolved, I would like to release v3.0.0 because:
- the fixes would take time
- the fixes will not change the APIs

**Comment by vdukhovni** [comment]
Seems largely fair.  A quick question.  How much work do you think it would take to enable EDNS0 support? My DANE scanner asks for DNSKEY records, which are invariably large enough to require TCP connections in the absence of EDNS0 support.  This would go much faster with EDNS0 and UDP.  Do you know how much work remains to enable EDNS0 in the client?  If not much remains to be done, perhaps that would a great final addition?

**Comment by kazu-yamamoto** [comment]
I guess that it's fairly easy.
Do you think that we can turn on EDNS0 on the first UDP try by default?

**Comment by vdukhovni** [comment]
Largely, yes, I just tried a crude patch, which works with my resolver, which unconditionally supports EDNS0:
```
diff --git a/Network/DNS/Encode.hs b/Network/DNS/Encode.hs
index 127cc13..8dac790 100644
--- a/Network/DNS/Encode.hs
+++ b/Network/DNS/Encode.hs
@@ -65,6 +65,7 @@ composeQueryAD idt qs = encode qry
               }
            }
         , question = qs
+       , additional = [fromEDNS0 defaultEDNS0]
         }
 
 ----------------------------------------------------------------
```

Which seems to work well for me.   More generally, we'd need to fall back to not using EDNS0 on  `FORMERR` replies to EDNS0, and possibly remember this for the each configured resolver (given a suitable MVar map or vector with a slot per peer nameserver).

I don't know whether we can easily support UDP EDNS0 buffer sizes above 4K, the Socket conduit used to read UDP data seems to do 4K reads, but UDP datagrams are truncated by recv calls, even if not read in full.  I don't recall whether conduit supports custom buffer sizes for UDP, but we should be willing to read up to the smaller of 64K (maximum UDP payload size) and the requested EDNS0 buffer size.

**Comment by vdukhovni** [comment]
The Socket conduit is really designed for TCP and has a hard-coded 4K buffer size:

https://hackage.haskell.org/package/network-conduit-0.3.0/docs/src/Data-Conduit-Network.html#sourceSocket
```
sourceSocket :: MonadIO m => Socket -> Source m ByteString
sourceSocket socket =
    src
  where
    src = SourceM pull close

    pull = do
        bs <- liftIO (recv socket 4096)
        return $ if S.null bs then Closed else Open src close bs
    close = return ()
```

**Comment by kazu-yamamoto** [comment]
@vdukhovni Do you want to support buffer size over 4096?

I have cleaned up the low level APIs on https://github.com/kazu-yamamoto/dns/tree/low-level-api

Unnecessary `ResourceT` has been removed. This would ease to switch from `conduit` to another streaming library if necessary.

**Comment by vdukhovni** [comment]
As for the low-level API, I'd need a few sentences about where you're going with this (the rationale for the changes and what the new design is supposed to do) before I can given useful feedback.  I see code moved around, but it is not immediately obvious what this means.

**Comment by kazu-yamamoto** [comment]
I will split issues raised here and record a new issue for each topic.

**Comment by kazu-yamamoto** [comment]
Are we ready for release of version 3.0.0?

**Comment by adinapoli-iohk** [comment]
@kazu-yamamoto No objection on my side, but I never had any to begin with, so I guess this is the least-interesting reply :)

**Comment by vdukhovni** [comment]
> @kazu-yamamoto Are we ready for release of version 3.0.0?

Probably, but see my pull request about removing minttl.

I also want to double check for any uncaught exceptions in the TCP code, I saw one uncaught connection-related exception in my DANE scanner recently, and it might have  been from the DNS lookup code.  I am fairly sure the SMTP code catches all network exceptions, but I'll check that too.  The scan in question missed one domain out of 4.8 million doing multiple lookups for most of the domains (DS, DNSKEY, MX, ...), and this was the first such error in many such scans, so the error is quite rare, but worth a look.

**Comment by kazu-yamamoto** [comment]
#101 has been merged.

**Comment by kazu-yamamoto** [comment]
> I saw one uncaught connection-related exception in my DANE scanner recently

Would you tell me the exact error type and value?

**Comment by vdukhovni** [comment]
> @kazu-yamamoto Would you tell me the exact error type and value?

I don't have the exact context saved.  All I remember is "resource vanished", which sounds like a connection reset or refused. The TCP code path seems to have a catcher for IO exceptions around it, I don't see any immediate issues...  If it happens again (might take a long time) I'll try to save more context...

**Comment by kazu-yamamoto** [comment]
>  All I remember is "resource vanished"

Very strange...

> If it happens again (might take a long time) I'll try to save more context...

OK. Thanks.

**Comment by kazu-yamamoto** [comment]
I have released v3.0.0.
Thank you for your contributions, all guys!

**Comment by adinapoli-iohk** [comment]
Smashing news, thanks @kazu-yamamoto !
